c/tests/test_stats.c

```diff
@@ -1044,6 +1068,52 @@ test_paper_ex_diversity(void)
     tsk_treeseq_free(&ts);
 }
 
+static void
+test_paper_ex_inheritance_covariance_errors(void)
+{
+    tsk_treeseq_t ts;
+
+    tsk_treeseq_from_text(&ts, 10, paper_ex_nodes, paper_ex_edges,
+            NULL, paper_ex_sites, paper_ex_mutations, paper_ex_individuals, NULL);
+    verify_one_way_weighted_func_errors(&ts, tsk_treeseq_inheritance_covariance);
+    tsk_treeseq_free(&ts);
+}
+
+static void
+test_paper_ex_inheritance_covariance(void)
+{
+    tsk_treeseq_t ts;
+    double result;
+    tsk_size_t num_samples;
+    double *weights;
+    size_t j;
+    int ret;
+
+    tsk_treeseq_from_text(&ts, 10, paper_ex_nodes, paper_ex_edges,
+            NULL, paper_ex_sites, paper_ex_mutations, paper_ex_individuals, NULL);
+    num_samples = tsk_treeseq_get_num_samples(&ts);
+    weights = malloc(num_samples * sizeof(double));
+    for (j = 0; j < num_samples; j++) {
+        weights[j] = 1.0;
+    }
+
+    ret = tsk_treeseq_inheritance_covariance(&ts, 1, weights, 0, NULL,
+            &result, TSK_STAT_SITE);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_DOUBLE_EQUAL_FATAL(result, 15.0, 1e-6);
+
+    /* weights of 0 leads to 0 */
+    for (j = 0; j < num_samples; j++) {
+        weights[j] = 0.0;
+    }
+    ret = tsk_treeseq_inheritance_covariance(&ts, 1, weights, 0, NULL,
+            &result, TSK_STAT_SITE);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_DOUBLE_EQUAL_FATAL(result, 0.0, 1e-6);
+
+    tsk_treeseq_free(&ts);
```

Comment:
Also leaking here. It's a good idea to run these tests through valgrind, catches these problems up front.